designsafe/apps/workspace/api/views.py

```diff
@@ -0,0 +1,293 @@
+    AppListingEntry,
+    AppTrayCategory,
+    AppVariant,
+)
+
+
+logger = logging.getLogger(__name__)
+METRICS = logging.getLogger("metrics.{}".format(__name__))
+
+
+def _app_license_type(app_def):
+    app_lic_type = getattr(app_def.notes, "licenseType", None)
+    lic_type = app_lic_type if app_lic_type in LICENSE_TYPES else None
+    return lic_type
+
+
+def _get_user_app_license(license_type, user):
+    _, license_models = get_license_info()
+    license_model = next(
+        (x for x in license_models if x.license_type == license_type), None
+    )
+    if not license_model:
+        return None
+    lic = license_model.objects.filter(user=user).first()
+    return lic
+
+
+def _get_app(app_id, app_version, user):
+    tapis = user.tapis_oauth.client
+    if app_version:
+        app_def = tapis.apps.getApp(appId=app_id, appVersion=app_version)
+    else:
+        app_def = tapis.apps.getAppLatestVersion(appId=app_id)
+    data = {"definition": app_def}
+
+    # GET EXECUTION SYSTEM INFO TO PROCESS SPECIFIC SYSTEM DATA E.G. QUEUE INFORMATION
+    data["exec_sys"] = tapis.systems.getSystem(
+        systemId=app_def.jobAttributes.execSystemId
+    )
+
+    lic_type = _app_license_type(app_def)
+    data["license"] = {"type": lic_type}
+    if lic_type is not None:
+        lic = _get_user_app_license(lic_type, user)
+        data["license"]["enabled"] = lic is not None
+
+    return data
+
+
+@method_decorator(login_required, name="dispatch")
+class AppsView(BaseApiView):
+    def get(self, request, *args, **kwargs):
+        tapis = request.user.tapis_oauth.client
+        app_id = request.GET.get("appId")
+        if app_id:
+            app_version = request.GET.get("appVersion")
+            METRICS.info(
+                "user:{} is requesting app id:{} version:{}".format(
+                    request.user.username, app_id, app_version
+                )
+            )
+            data = _get_app(app_id, app_version, request.user)
+
+            # Check if default storage system needs keys pushed
+            if settings.PORTAL_DATAFILES_DEFAULT_STORAGE_SYSTEM:
+                system_id = settings.PORTAL_DATAFILES_DEFAULT_STORAGE_SYSTEM["system"]
+                system_def = tapis.systems.getSystem(systemId=system_id)
+
+                try:
+                    tapis.files.listFiles(systemId=system_id, path="/")
+                except (InternalServerError, UnauthorizedError):
+                    data["systemNeedsKeys"] = True
+                    data["pushKeysSystem"] = system_def
+        else:
+            METRICS.info("user:{} is requesting all apps".format(request.user.username))
+            data = {"appListing": tapis.apps.getApps()}
+
+        return JsonResponse(
+            {
+                "status": 200,
+                "response": data,
+            },
+            encoder=BaseTapisResultSerializer,
+        )
+
+
+@method_decorator(login_required, name="dispatch")
```

Comment:
`@login_required` can cause weird redirect behavior, since it's designed for HTML template views. I think we should return a JSON response with a 401 code if the user isn't authenticated.

Reply:
doesn't `login_required` redirect to login? Is that not desired behavior?

Reply:
It's desired for an HTML view, but for a JSON endpoint it causes weird behaviors like redirecting to a blob of raw JSON after authentication.

Reply:
Hmm I guess I'm not following. Isn't this going to be a template view?

Reply:
Alright, added a new class `AuthenticatedApiView` which just returns that 401. Lmk what you think
![Screenshot 2024-03-07 at 6 22 56 PM](https://github.com/DesignSafe-CI/portal/assets/20326896/96a63609-cf14-4b6d-975d-be799d4329c3)